alf/algorithms/agent.py

```diff
@@ -101,32 +104,31 @@ def __init__(self,
             debug_summaries (bool): True if debug summaries should be created.
             name (str): Name of this algorithm.
             """
-        train_state_spec = AgentState()
-        predict_state_spec = AgentState()
-
-        if intrinsic_curiosity_module is not None:
-            train_state_spec = train_state_spec._replace(
-                icm=intrinsic_curiosity_module.train_state_spec)
+        state_specs = AgentStateSpecs()
 
+        ## 1. goal generator
         rl_observation_spec = observation_spec
         if goal_generator is not None:
-            train_state_spec = train_state_spec._replace(
-                goal_generator=goal_generator.train_state_spec)
-            predict_state_spec = predict_state_spec._replace(
-                goal_generator=goal_generator.predict_state_spec)
+            state_specs.collect_algorithm_state_specs(goal_generator,
+                                                      "goal_generator")
             rl_observation_spec = [
                 observation_spec, goal_generator.action_spec
             ]
 
+        ## 2. rl algorithm
         rl_algorithm = rl_algorithm_cls(
             observation_spec=rl_observation_spec,
             action_spec=action_spec,
+            env=env,
```

Comment:
no need to provide `env`? It's not root algorithm.

Reply:
yeah that's a bug fixed locally